# Project 8 -- Niharika Rai

**TA Help:** (for instance) John Smith, Alice Jones, etc., list names of any TAs who helped you

- For example: Help with figuring out how to write a function (describe the tasks that they helped you with)

**Collaboration:** My Friend in CS, My Uncle, Another Student, etc., list names of any other people who helped you

(describe the tasks that they helped you with)
- For example: helped figuring out how to load the dataset.
- Another example: helped debug error with my plot.

**Internet Resources:** Stack Exchange, Stack Overflow, etc.

(describe any information that you learned from internet resources, including the URLs)
- data frames in Pandas versus R from StackOverflow  https://stackoverflow.com/questions/8991709/why-were-pandas-merges-in-python-faster-than-data-table-merges-in-r-in-2012

**ChatGPT, Gemini, Claude, etc:** Any language models or generative AI chatbots that helped you.

(if you used any such tools, please tell us here)
- For example:  I asked ChatGPT how to define a new data frames
- Another example:  Gemini told me how to make a function for sorting my data

- ***Link to AI Chat History***: Here's the link: https://chatgpt.com/share/68f56b01-4470-800d-beb4-6b6746879589. Please share a link to your chat if you used AI (ex. ChatGPT Shared Links)
**OVERALL MESSAGE:** Any time that you used anything except your brain to solve the questions in these projects, you need to disclose such resources at the start of the project, with details about your usage of the tools.

**YOUR OWN WORK:** Even when you utilize other resources, do NOT just copy and paste.  Write all explanations in your own words, using several sentences in English, which are understandable and which you wrote (and did not just copy and paste).

## Question 1

In [62]:
import sqlite3

# Load the Lahman database
read_only_lahman = sqlite3.connect('/anvil/projects/tdm/data/lahman/lahman.db')
local_lahman = sqlite3.connect('./lahman.db')

read_only_lahman.backup(local_lahman) # Backup the database to your local file
read_only_lahman.close()
local_lahman.close()

In [63]:
%sql sqlite:///./lahman.db

In [64]:
%%sql
SELECT name FROM sqlite_master WHERE type='table';

 * sqlite:///./lahman.db
Done.


name
teamsfranchises
teams
teamshalf
people
allstarfull
appearances
awardsmanagers
awardsplayers
awardssharemanagers
awardsshareplayers


In [65]:
%%sql
SELECT * FROM Teams LIMIT 5;

 * sqlite:///./lahman.db
Done.


yearID,lgID,teamID,franchID,divID,teamRank,G,Ghome,W,L,DivWin,WCWin,LgWin,WSWin,R,AB,H,2B,3B,HR,BB,SO,SB,CS,HBP,SF,RA,ER,ERA,CG,SHO,SV,IPouts,HA,HRA,BBA,SOA,E,DP,FP,name,park,attendance,BPF,PPF,teamIDBR,teamIDlahman45,teamIDretro
1871,NA,BS1,BNA,,3,31,,20,10,,,N,,401,1372,426,70,37,3,60,19,73,16,,,303,109,3.55,22,1,3,828,367,2,42,23,243,24,0.834,Boston Red Stockings,South End Grounds I,,103,98,BOS,BS1,BS1
1871,NA,CH1,CNA,,2,28,,19,9,,,N,,302,1196,323,52,21,10,60,22,69,21,,,241,77,2.76,25,0,1,753,308,6,28,22,229,16,0.829,Chicago White Stockings,Union Base-Ball Grounds,,104,102,CHI,CH1,CH1
1871,NA,CL1,CFC,,8,29,,10,19,,,N,,249,1186,328,35,40,7,26,25,18,8,,,341,116,4.11,23,0,0,762,346,13,53,34,234,15,0.818,Cleveland Forest Citys,National Association Grounds,,96,100,CLE,CL1,CL1
1871,NA,FW1,KEK,,7,19,,7,12,,,N,,137,746,178,19,8,2,33,9,16,4,,,243,97,5.17,19,1,0,507,261,5,21,17,163,8,0.803,Fort Wayne Kekiongas,Hamilton Field,,101,107,KEK,FW1,FW1
1871,NA,NY2,NNA,,5,33,,16,17,,,N,,302,1404,403,43,21,1,33,15,46,15,,,313,121,3.72,32,1,0,879,373,7,42,22,235,14,0.84,New York Mutuals,Union Grounds (Brooklyn),,90,88,NYU,NY2,NY2


1.1 I copied the Lahman database to my local folder in anvil. I used the sample code given in the question to achieve this.  
1.2 I loaded the first 5 rows of the Teams table. I used the sample code given in the question to achieve this. 

## Question 2

In [66]:
%%sql 
CREATE TABLE MyPlayers (
    playerID TEXT PRIMARY KEY,
    name TEXT NOT NULL,
    teamID TEXT NOT NULL
);

 * sqlite:///./lahman.db
Done.


[]

In [67]:
%%sql
INSERT INTO MyPlayers (playerID, name, teamID)
VALUES ('player1', 'John Doe', 'team1');

 * sqlite:///./lahman.db
1 rows affected.


[]

In [68]:
%%sql
SELECT * FROM MyPlayers;

 * sqlite:///./lahman.db
Done.


playerID,name,teamID
player1,John Doe,team1


In [69]:
%%sql
INSERT INTO MyPlayers (playerID, name, teamID)
VALUES
('player2', 'Jane Smith', 'team1'),
('player3', 'Bob Johnson', 'team2'),
('player4', 'Alice Brown', 'team2');

 * sqlite:///./lahman.db
3 rows affected.


[]

In [70]:
%%sql
SELECT * FROM MyPlayers;

 * sqlite:///./lahman.db
Done.


playerID,name,teamID
player1,John Doe,team1
player2,Jane Smith,team1
player3,Bob Johnson,team2
player4,Alice Brown,team2


In [71]:
%%sql 
INSERT INTO MyPlayers (playerID, name, teamID)
SELECT p.playerID, p.nameFirst || ' ' || p.nameLast AS name, MIN(s.teamID) as teamID
FROM People p
JOIN Salaries s on s.playerID = p.playerID
WHERE s.yearID = ( SELECT MAX(yearID) FROM Salaries s2 WHERE s2.playerID = s.playerID )
GROUP BY p.playerID;

 * sqlite:///./lahman.db
5149 rows affected.


[]

In [72]:
%%sql
SELECT * FROM MyPlayers LIMIT 5;

 * sqlite:///./lahman.db
Done.


playerID,name,teamID
player1,John Doe,team1
player2,Jane Smith,team1
player3,Bob Johnson,team2
player4,Alice Brown,team2
aardsda01,David Aardsma,NYA


2.1 I used the sample code from the question. 
2.2 I used the sample code from the question. 
2.3 I used the sample code from the question. 
2.4 There was no 'Masters' table in the Lahman database. I referred to Piazza discussion board and found that this is a typo. It was actually the 'People' table and was able to use the sample code from the question to insert all players. 
2.5 I displayed the first 5 rows of 'MyPlayers' table. The fifth row indicates that players from the 'People' table have been added as well.

## Question 3

In [73]:
%%sql 
ALTER TABLE MyPlayers
ADD COLUMN height REAL;

 * sqlite:///./lahman.db
Done.


[]

In [74]:
%%sql
SELECT * FROM MyPlayers LIMIT 5;

 * sqlite:///./lahman.db
Done.


playerID,name,teamID,height
player1,John Doe,team1,None
player2,Jane Smith,team1,None
player3,Bob Johnson,team2,None
player4,Alice Brown,team2,None
aardsda01,David Aardsma,NYA,None


In [75]:
%%sql 
ALTER TABLE MyPlayers
ADD COLUMN weight REAL;

 * sqlite:///./lahman.db
Done.


[]

In [76]:
%%sql
SELECT * FROM MyPlayers LIMIT 5;

 * sqlite:///./lahman.db
Done.


playerID,name,teamID,height,weight
player1,John Doe,team1,None,None
player2,Jane Smith,team1,None,None
player3,Bob Johnson,team2,None,None
player4,Alice Brown,team2,None,None
aardsda01,David Aardsma,NYA,None,None


In [77]:
%%sql 
ALTER TABLE MyPlayers
ADD COLUMN age INTEGER;

 * sqlite:///./lahman.db
Done.


[]

In [78]:
%%sql
SELECT * FROM MyPlayers LIMIT 5;

 * sqlite:///./lahman.db
Done.


playerID,name,teamID,height,weight,age
player1,John Doe,team1,None,None,None
player2,Jane Smith,team1,None,None,None
player3,Bob Johnson,team2,None,None,None
player4,Alice Brown,team2,None,None,None
aardsda01,David Aardsma,NYA,None,None,None


In [79]:
%%sql
ALTER TABLE MyPlayers
DROP COLUMN teamID;

 * sqlite:///./lahman.db
Done.


[]

In [80]:
%%sql
SELECT * FROM MyPlayers LIMIT 5;

 * sqlite:///./lahman.db
Done.


playerID,name,height,weight,age
player1,John Doe,None,None,None
player2,Jane Smith,None,None,None
player3,Bob Johnson,None,None,None
player4,Alice Brown,None,None,None
aardsda01,David Aardsma,None,None,None


In [81]:
%%sql
INSERT INTO MyPlayers (playerID, name, height, weight, age)
VALUES
('player5', 'Sam Stevens', 177, 180, 25),
('player6', 'Barbara Handler', 175, 160, 24),
('player7', 'William Bryson', 185, 170, 26),
('player8', 'Harry Chancellor', 188, 185, 27),
('player9', 'Sofie Strong', 176, 165, 28);

 * sqlite:///./lahman.db
5 rows affected.


[]

In [82]:
%%sql 
SELECT * FROM MyPlayers 
ORDER BY height DESC
LIMIT 5;

 * sqlite:///./lahman.db
Done.


playerID,name,height,weight,age
player8,Harry Chancellor,188.0,185.0,27
player7,William Bryson,185.0,170.0,26
player5,Sam Stevens,177.0,180.0,25
player9,Sofie Strong,176.0,165.0,28
player6,Barbara Handler,175.0,160.0,24


3.1 I used the sample query from the question but modified it to 'height' as the new column being added and 'REAL' as its column type. 
3.2 I used the sample query from the question but modified it to 'weight' as the new column being added and 'REAL' as its column type.
3.3 I first added the age column and dropped the teamID column based on the sample queries given in the question. I, then, inserted 5 example players, giving them each an unique playerID, name, height, weight and age. I used the 'INSERT INTO' command based on the previous question's sample query. 
3.4 I selected all the rows from 'MyPlayers' table, ordered it by height being in descending order and limited the rows to 5, to get the top 5 tallest players, ordered by height in descending order. 

## Question 4

In [83]:
%%sql 
DELETE FROM MyPlayers
WHERE playerID = 'player1';

 * sqlite:///./lahman.db
1 rows affected.


[]

In [84]:
%%sql
SELECT * FROM MyPlayers LIMIT 5;

 * sqlite:///./lahman.db
Done.


playerID,name,height,weight,age
player2,Jane Smith,None,None,None
player3,Bob Johnson,None,None,None
player4,Alice Brown,None,None,None
aardsda01,David Aardsma,None,None,None
aasedo01,Don Aase,None,None,None


In [85]:
%%sql
DELETE FROM MyPlayers
WHERE height < 182.88;

 * sqlite:///./lahman.db
3 rows affected.


[]

In [86]:
%%sql 
SELECT * FROM MyPlayers 
ORDER BY height DESC
LIMIT 5;

 * sqlite:///./lahman.db
Done.


playerID,name,height,weight,age
player8,Harry Chancellor,188.0,185.0,27
player7,William Bryson,185.0,170.0,26
player2,Jane Smith,None,None,None
player3,Bob Johnson,None,None,None
player4,Alice Brown,None,None,None


4.1 I deleted a player from the 'MyPlayers' table using the 'DELETE' command. I used the sample query given in the question to achieve this. I also viewed the first 5 rows from 'MyPlayers' table and 'player1' did not show up because that is the playerID value which I deleted from the table.   
4.2 I deleted all players from the 'MyPlayers' table that have a height less than 6 feet using the 'DELETE' command. I used the sample query given in the question but modified it based on my height value units. I entered the heights in centimeters. So deleting all heights less than 6 feet would mean deleting all heights that are less than 182.88 cm. This is because 6 feet = 182.88 cm. Running this command deleted 3 players from 'MyPlayers' table as they were less than 6 feet tall.    
4.3 I queried 'MyPlayers' table to verify that the players have been deleted by again sorting the heights in descending order and limiting the rows to 5. Hence, the players who were less than 6 feet tall did not show in the query output. 

## Question 5

In [90]:
%%sql
TRUNCATE TABLE MyPlayers;

 * sqlite:///./lahman.db
(sqlite3.OperationalError) near "TRUNCATE": syntax error
[SQL: TRUNCATE TABLE MyPlayers;]
(Background on this error at: https://sqlalche.me/e/20/e3q8)


In [91]:
%%sql 
DELETE FROM MyPlayers;

 * sqlite:///./lahman.db
5154 rows affected.


[]

In [92]:
%%sql
SELECT * FROM MyPlayers LIMIT 5;

 * sqlite:///./lahman.db
Done.


playerID,name,height,weight,age


In [93]:
%%sql
DROP TABLE MyPlayers;

 * sqlite:///./lahman.db
Done.


[]

In [94]:
%%sql
SELECT * FROM MyPlayers LIMIT 5;

 * sqlite:///./lahman.db
(sqlite3.OperationalError) no such table: MyPlayers
[SQL: SELECT * FROM MyPlayers LIMIT 5;]
(Background on this error at: https://sqlalche.me/e/20/e3q8)


5.1 I used the sample query given in the question to achieve this. The question says to use the 'TRUNCATE' command but the sample query used the 'DELETE' command so I used that instead. I tried experimenting first with 'TRUNCATE TABLE' command (found the syntax using ChatGPT) as well but that did not work either so I used the sample query given in the question instead.  
5.2 I used the sample query given in the question to achieve this. 
5.3 I used the sample query given in the question to verify that 'MyPlayers' table does not exist anymore.

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.

https://www.purdue.edu/odos/osrr/honor-pledge/
